mujoco_warp/_src/io.py

```diff
@@ -1649,3 +1653,202 @@ def get_data_into(
+  qacc_warmstart_out: wp.array2d(dtype=float),
+  qfrc_applied_out: wp.array2d(dtype=float),
+  qacc_out: wp.array2d(dtype=float),
+):
+  worldid, dofid = wp.tid()
+  qvel_out[worldid, dofid] = 0.0
+  qacc_warmstart_out[worldid, dofid] = 0.0
+  qfrc_applied_out[worldid, dofid] = 0.0
+  qacc_out[worldid, dofid] = 0.0
+
+
+@wp.kernel
+def _zero_nu(
+  # Model:
+  na: int,
+  # Data out:
+  act_out: wp.array2d(dtype=float),
+  ctrl_out: wp.array2d(dtype=float),
+  act_dot_out: wp.array2d(dtype=float),
+):
+  worldid, actid = wp.tid()
+  ctrl_out[worldid, actid] = 0.0
+  if actid < na:
+    act_out[worldid, actid] = 0.0
+    act_dot_out[worldid, actid] = 0.0
+
+
+@wp.kernel
+def _set_mocap(
+  # Model:
+  body_mocapid: wp.array(dtype=int),
+  body_pos: wp.array2d(dtype=wp.vec3),
+  body_quat: wp.array2d(dtype=wp.quat),
+  # Data out:
+  mocap_pos_out: wp.array2d(dtype=wp.vec3),
+  mocap_quat_out: wp.array2d(dtype=wp.quat),
+):
+  worldid, bodyid = wp.tid()
+
+  mocapid = body_mocapid[bodyid]
+
+  if mocapid >= 0:
+    mocap_pos_out[worldid, mocapid] = body_pos[worldid, bodyid]
+    mocap_quat_out[worldid, mocapid] = body_quat[worldid, bodyid]
+
+
+@wp.kernel
+def _zero_contact(
+  # Data in:
+  ncon_in: wp.array(dtype=int),
+  # In:
+  nefcaddress: int,
+  # Data out:
+  contact_dist_out: wp.array(dtype=float),
+  contact_pos_out: wp.array(dtype=wp.vec3),
+  contact_frame_out: wp.array(dtype=wp.mat33),
+  contact_includemargin_out: wp.array(dtype=float),
+  contact_friction_out: wp.array(dtype=types.vec5),
+  contact_solref_out: wp.array(dtype=wp.vec2),
+  contact_solreffriction_out: wp.array(dtype=wp.vec2),
+  contact_solimp_out: wp.array(dtype=types.vec5),
+  contact_dim_out: wp.array(dtype=int),
+  contact_geom_out: wp.array(dtype=wp.vec2i),
+  contact_efc_address_out: wp.array2d(dtype=int),
+  contact_worldid_out: wp.array(dtype=int),
+):
+  conid = wp.tid()
+
+  if conid >= ncon_in[0]:
+    return
+
+  contact_dist_out[conid] = 0.0
+  contact_pos_out[conid] = wp.vec3(0.0)
+  contact_frame_out[conid] = wp.mat33(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_includemargin_out[conid] = 0.0
+  contact_friction_out[conid] = types.vec5(0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_solref_out[conid] = wp.vec2(0.0, 0.0)
+  contact_solreffriction_out[conid] = wp.vec2(0.0, 0.0)
+  contact_solimp_out[conid] = types.vec5(0.0, 0.0, 0.0, 0.0, 0.0)
+  contact_dim_out[conid] = 0
+  contact_geom_out[conid] = wp.vec2i(0, 0)
+  for i in range(nefcaddress):
+    contact_efc_address_out[conid, i] = 0
+  contact_worldid_out[conid] = 0
+
+
+def reset_data(m: types.Model, d: types.Data):
+  """Clear data, set defaults."""
+  # copy qpos0 from model
+  wp.copy(d.qpos, m.qpos0)
+
+  # # zero out arrays that are not affected by forward
+  wp.launch(_zero_nv, dim=(d.nworld, m.nv), outputs=[d.qvel, d.qacc_warmstart, d.qfrc_applied, d.qacc])
+  wp.launch(_zero_nu, dim=(d.nworld, m.nu), inputs=[m.na], outputs=[d.act, d.ctrl, d.act_dot])
+  wp.copy(d.eq_active, m.eq_active0)
```

Comment:
hmmm... same question here, in `make_data` we explicitly tile this, so it repeats across all worlds... note that this field is NOT batched at all in `mujoco_warp` Model

are we sure this line is doing the right thing?

Reply:
updated to set this field in `_set_nworld`: `eq_active[worldid, i] = eq_active0[i]`